randao/src/Randao.sol

```diff
@@ -0,0 +1,72 @@
+// SPDX-License-Identifier: MIT
+// WARNING! This smart contract and the associated zk-SNARK verifiers have not been audited.
+// DO NOT USE THIS CONTRACT FOR PRODUCTION
+pragma solidity ^0.8.12;
+
+import {IAxiomV1Query} from "axiom-contracts/contracts/interfaces/IAxiomV1Query.sol";
+import {Ownable} from "openzeppelin-contracts/access/Ownable.sol";
+import {RLPReader} from "utils/RLPReader.sol";
+
+contract Randao is Ownable {
+    using RLPReader for RLPReader.RLPItem;
+    using RLPReader for bytes;
+
+    address public axiomQueryAddress;
+    uint32 MERGE_BLOCK = 15537393;
+
+    // mapping between blockNumber and prevRandao
+    mapping(uint32 => uint256) public prevRandaos;
+
+    event RandaoProof(uint32 blockNumber, uint256 prevRandao);
+
+    event UpdateAxiomQueryAddress(address newAddress);
+
+    constructor(address _axiomQueryAddress) {
+        axiomQueryAddress = _axiomQueryAddress;
+    }
+
+    function updateAxiomAddress(address _axiomQueryAddress) external onlyOwner {
+        axiomQueryAddress = _axiomQueryAddress;
+        emit UpdateAxiomQueryAddress(_axiomQueryAddress);
+    }
+
+    function _getRandaoFromBlock(
+        uint32 blockNumber,
+        bytes32 blockHash,
+        bytes memory rlpEncodedHeader
+    ) internal pure returns (uint256) {
+        require(keccak256(rlpEncodedHeader) == blockHash, "invalid blockhash");
+
+        RLPReader.RLPItem[] memory ls = rlpEncodedHeader.toRlpItem().toList();
+        require(blockNumber == ls[8].toUint(), "invalid block number");
+        uint256 randao = ls[13].toUint();
+        return randao;
+    }
+
+    function verifyRandao(
```

Comment:
This can work in the same way as before, so no query is needed -- can directly read from `IAxiomV1Verifier` using `isBlockHashValid` and `isRecentBlockHashValid`: https://github.com/axiom-crypto/axiom-apps/blob/main/randao/contracts/src/RandaoV02.sol#L35